File: typescript/exercises.ts
import { open } from "node:fs/promises"

export function change(amount: bigint): Map<bigint, bigint> {
  if (amount < 0) {
    throw new RangeError("Amount cannot be negative")
  }
  let counts: Map<bigint, bigint> = new Map()
  let remaining = amount
  for (const denomination of [25n, 10n, 5n, 1n]) {
    counts.set(denomination, remaining / denomination)
    remaining %= denomination
  }
  return counts
}

export function firstThenApply<T, U>(arr: T[], predicate: (x: T) => boolean, func: (x: T) => U): U | undefined {
  const result = arr.find(predicate);
  return result !== undefined ? func(result) : undefined;
}

export function* powersGenerator(base: bigint): Generator<bigint> {
  let result = 1n;
  while (true) {
      yield result;
      result *= base;
  }
}

export async function meaningfulLineCount(filePath: string): Promise<number> {
    const fileHandle = await open(filePath, 'r');
    try {
        const fileContent: string = await fileHandle.readFile('utf-8');
        const lines: string[] = fileContent.split('\n');
        return lines.filter((line: string) => {
            const trimmed: string = line.trim();
            return trimmed.length > 0 && !trimmed.startsWith('#');
        }).length;
    } finally {
        await fileHandle.close();
    }
}

export type Shape = 
  { kind: "Box", width: number, length: number, depth: number } | 
  { kind: "Sphere", radius: number };

export function surfaceArea(shape: Shape): number {
    if (shape.kind === "Box") {
        return 2 * (shape.width * shape.length + shape.length * shape.depth + shape.width * shape.depth);
    } else {
        return 4 * Math.PI * shape.radius ** 2;
    }
}

export function volume(shape: Shape): number {
    if (shape.kind === "Box") {
        return shape.width * shape.length * shape.depth;
    } else {
        return (4 / 3) * Math.PI * shape.radius ** 3;
    }
}

export interface BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T>;
  contains(value: T): boolean;
  size(): number;
  inorder(): Generator<T>;
  toString(): string;
}

export class Empty<T> implements BinarySearchTree<T> {
  insert(value: T): BinarySearchTree<T> {
      return new Node<T>(value, new Empty<T>(), new Empty<T>());
  }

  contains(_: T): boolean {
      return false;
  }

  size(): number {
      return 0;
  }

  *inorder(): Generator<T> {}

  toString(): string {
      return "()";
  }
}

export class Node<T> implements BinarySearchTree<T> {
  constructor(
      private readonly value: T,
      private readonly left: BinarySearchTree<T>,
      private readonly right: BinarySearchTree<T>
  ) {}

  insert(value: T): BinarySearchTree<T> {
    if (this instanceof Empty) {
      return new Node(value, new Empty(), new Empty());
    }
    
    if (value < this.value) {
      return new Node(this.value, this.left.insert(value), this.right); // Return new tree with updated left
    } else if (value > this.value) {
      return new Node(this.value, this.left, this.right.insert(value)); // Return new tree with updated right
    } else {
      return this; // If the value is already in the tree, return the same tree
    }
  }

  contains(value: T): boolean {
      if (value === this.value) {
          return true;
      } else if (value < this.value) {
          return this.left.contains(value);
      } else {
          return this.right.contains(value);
      }
  }

  size(): number {
      return 1 + this.left.size() + this.right.size();
  }

  *inorder(): Generator<T> {
    if (!(this instanceof Empty)) {
      yield* this.left.inorder();  // Traverse the left subtree
      yield this.value;            // Yield the current node's value
      yield* this.right.inorder(); // Traverse the right subtree
    }
  }

  toString(): string {
    const leftStr = this.left instanceof Empty ? "" : this.left.toString();
    const rightStr = this.right instanceof Empty ? "" : this.right.toString();
    return `(${leftStr}${this.value}${rightStr})`;
  }
}

File: ocaml/exercises.ml
exception Negative_Amount

let change amount =
  if amount < 0 then
    raise Negative_Amount
  else
    let denominations = [25; 10; 5; 1] in
    let rec aux remaining denominations =
      match denominations with
      | [] -> []
      | d :: ds -> (remaining / d) :: aux (remaining mod d) ds
    in
    aux amount denominations

let rec first_then_apply (lst: 'a list) (predicate: 'a -> bool) (func: 'a -> 'b option): 'b option =
  match lst with
  | [] -> None
  | x :: xs -> 
    if predicate x then (func x)
    else first_then_apply xs predicate func

let rec powers_generator (base: int) : int Seq.t =
  let rec aux (current: int) () = Seq.Cons (current, aux (current * base)) in
  aux 1
  

let meaningful_line_count (file: string) : int =
  let ic = open_in file in
  Fun.protect
    ~finally:(fun () -> close_in ic)
    (fun () ->
      let rec count (acc: int) : int =
        try
          let line = input_line ic in
          let trimmed = String.trim line in
          if String.length trimmed > 0 && not (String.sub trimmed 0 1 = "#") then
            count (acc + 1)
          else
            count acc
        with End_of_file -> acc
      in count 0)


type shape =
  | Sphere of float
  | Box of float * float * float

let volume (s: shape) : float =
  match s with
  | Sphere radius -> (4.0 /. 3.0) *. Float.pi *. radius ** 3.0
  | Box (width, height, depth) -> width *. height *. depth


let surface_area (s: shape) : float =
  match s with
  | Sphere radius -> 4.0 *. Float.pi *. radius ** 2.0
  | Box (width, height, depth) -> 2.0 *. (width *. height +. height *. depth +. depth *. width)


let to_string (s: shape) : string =
  match s with
  | Sphere radius -> Printf.sprintf "Sphere with radius: %.2f" radius
  | Box (width, height, depth) ->
      Printf.sprintf "Box with width: %.2f, height: %.2f, depth: %.2f" width height depth


module BST : sig
  type 'a bst

  val empty : 'a bst
  val insert : 'a -> 'a bst -> 'a bst
  val contains : 'a -> 'a bst -> bool
  val size : 'a bst -> int
  val inorder : 'a bst -> 'a list
end = struct

  type 'a bst = 
    | Empty
    | Node of 'a * 'a bst * 'a bst

  let empty = Empty

  let rec insert (x: 'a) (tree: 'a bst) : 'a bst =
    match tree with
    | Empty -> Node (x, Empty, Empty)
    | Node (v, left, right) ->
      if x < v then Node (v, insert x left, right)
      else if x > v then Node (v, left, insert x right)
      else tree

  let rec contains (x: 'a) (tree: 'a bst) : bool =
    match tree with
    | Empty -> false
    | Node (v, left, right) ->
      if x = v then true
      else if x < v then contains x left
      else contains x right

  let rec size (tree: 'a bst) : int =
    match tree with
    | Empty -> 0
    | Node (_, left, right) -> 1 + size left + size right

  let rec inorder (tree: 'a bst) : 'a list =
    match tree with
    | Empty -> []
    | Node (v, left, right) -> inorder left @ [v] @ inorder right
end

open BST
